internal/txmanager/txmanager_test.go

```diff
@@ -0,0 +1,253 @@
+}
+
+func (s *TxManagerTestSuite) TestManualSQL_Success() {
+	_, err := s.db.ExecContext(s.ctx, `
+        CREATE TABLE IF NOT EXISTS dummy (
+            id TEXT PRIMARY KEY,
+            value TEXT NOT NULL
+        );
+    `)
+	s.Require().NoError(err)
+
+	_, err = s.db.ExecContext(s.ctx, `DELETE FROM dummy;`)
+	s.Require().NoError(err)
+
+	err = s.txMgr.WithTransaction(s.ctx, func(ctx context.Context) error {
+		tx, err := txmanager.GetTx(ctx)
+		s.Require().NoError(err)
+		_, err = tx.ExecContext(ctx, `INSERT INTO dummy (id, value) VALUES ('id1', 'value1')`)
+		return err
+	})
+	s.Require().NoError(err)
+
+	var value string
+	err = s.db.GetContext(s.ctx, &value, "SELECT value FROM dummy WHERE id = $1", "id1")
+	s.Require().NoError(err)
+	s.Equal("value1", value)
+}
+
+func (s *TxManagerTestSuite) TestParallelTransactions_NonConflicting() {
+	_, err := s.db.ExecContext(s.ctx, `
+        CREATE TABLE IF NOT EXISTS dummy (
+            id TEXT PRIMARY KEY,
+            value TEXT NOT NULL
+        );
+    `)
+	s.Require().NoError(err)
+
+	_, err = s.db.ExecContext(s.ctx, `DELETE FROM dummy;`)
+	s.Require().NoError(err)
+
+	_, err = s.db.ExecContext(s.ctx, `
+        INSERT INTO dummy (id, value) VALUES ('id1', 'init'), ('id2', 'init');
+    `)
+	s.Require().NoError(err)
+
+	ch1 := make(chan struct{})
+	ch2 := make(chan struct{})
+
+	go func() {
+		err := s.txMgr.WithTransaction(s.ctx, func(ctx context.Context) error {
+			tx, err := txmanager.GetTx(ctx)
+			s.Require().NoError(err)
+
+			_, err = tx.ExecContext(ctx, `UPDATE dummy SET value = 'tx1' WHERE id = 'id1'`)
+			s.Require().NoError(err)
+
+			ch1 <- struct{}{}
+			<-ch2
+
+			return nil
+		})
+		s.Require().NoError(err)
+	}()
+
+	go func() {
+		err := s.txMgr.WithTransaction(s.ctx, func(ctx context.Context) error {
+			tx, err := txmanager.GetTx(ctx)
+			s.Require().NoError(err)
+
+			<-ch1
+			_, err = tx.ExecContext(ctx, `UPDATE dummy SET value = 'tx2' WHERE id = 'id2'`)
+			s.Require().NoError(err)
+
+			ch2 <- struct{}{}
+			return nil
+		})
+		s.Require().NoError(err)
+	}()
+
+	time.Sleep(2 * time.Second)
+
+	var v1, v2 string
+	s.Require().NoError(s.db.Get(&v1, `SELECT value FROM dummy WHERE id = 'id1'`))
+	s.Require().NoError(s.db.Get(&v2, `SELECT value FROM dummy WHERE id = 'id2'`))
+	s.Equal("tx1", v1)
+	s.Equal("tx2", v2)
+}
+
+func (s *TxManagerTestSuite) TestParallelTransactions_Conflicting() {
+	_, err := s.db.ExecContext(s.ctx, `
+        CREATE TABLE IF NOT EXISTS dummy (
+            id TEXT PRIMARY KEY,
+            value TEXT NOT NULL
+        );
+    `)
+	s.Require().NoError(err)
+
+	_, err = s.db.ExecContext(s.ctx, `DELETE FROM dummy;`)
+	s.Require().NoError(err)
+
+	_, err = s.db.ExecContext(s.ctx, `INSERT INTO dummy (id, value) VALUES ('id1', 'init');`)
+	s.Require().NoError(err)
+
+	chReady := make(chan struct{})
+	chContinue := make(chan struct{})
+	errCh := make(chan error, 2)
+
+	txOpts := &sql.TxOptions{Isolation: sql.LevelSerializable}
```

Comment:
BTW it should be a default options to use in WithTransaction, you do not need WithTransactionWithOptions specific method
it is usually a default in application and you rarely need any other isolation level unless your application has high load